test/plugin/scenarios/retransform-class-tomcat-scenario/configuration.yml

```diff
@@ -17,7 +17,5 @@
 type: tomcat
 entryService: http://localhost:8080/retransform-class-tomcat-scenario/case/retransform-class
 healthCheck: http://localhost:8080/retransform-class-tomcat-scenario/case/healthCheck
-environment:
-  - CATALINA_OPTS="-Dskywalking.agent.is_cache_enhanced_class=true -Dskywalking.agent.class_cache_mode=FILE"
 dependencies:
-withPlugins: apm-spring-annotation-plugin-*.jar
+withPlugins: apm-*.jar
```

Comment:
With all plugins? Are these necessary? I prefer the explicit names of plugins.

Reply:
retransform class test scenario should test with more plugins, especially with the jdk plugins, test scenario is not enough currently, may add more cases later.

Reply:
We should separate them, otherwise, it would be confusing when it fails.

Reply:
I think it is ok to load all plugins in this test scenario. In order to ensure the retransform function, plugins that affect retransform need to be identified in time. Retransform class is not a standalone feature and in principle all plugins need to support it.

Reply:
The point is not about these two plugin tests, it is about the stability of the testing. The plugins are going to be added/removed from time to time, if you copy all, actually, you don't know which plugins exist, which may cause these tests to fail by other contributions.

Reply:
I got it. It's better to test retransform in test scenario of each plugin if we think of retransform as a basic feature. 
Another point, there may be a conflict when some plugins put together, may be the class/method/field name generated by  class enhancement , how to find this out?

Reply:
I don't think general plugins matters. After all, the worst case, some plugins would work in retransfer case, right?

We could pick some for tests.
1. General Spring with another agents.
2. Tomcat(another class loader mode) with another agent
3. SkyWalking before or after another agents.
4. SkyWalking before or after bytebuddy based agent(a test agent maybe?)

Reply:
The worst case is not that some plugins don't support retransform, but that the plugins don't work together if there really is a class/method/field name conflict. So before I want to test with all the plugins, on you decide.

Reply:
I think we don't cover multiple plugins for one method case, how could that happens?

Reply:
And even yoi copied all, it doesn't approve this, because many plugins are not activated(not used) in your scenarios.

Reply:
> I think we don't cover multiple plugins for one method case, how could that happens?

We now support multiple plugins enhancing the same class, although enhancing different methods, but if the same method/field name is generated, there will be an error.

> And even yoi copied all, it doesn't verify this, because many plugins are not activated(not used) in your scenarios.

Yes, and we may need more testcases.

Reply:
> We now support multiple plugins enhancing the same class, although enhancing different methods, but if the same method/field name is generated, there will be an error.

Yes, but that could be error when the plugin tests run too. What your new cases cover is only re-transfer. They are verifying for different cases.

Reply:
It's not that simple. Different plugin enhancers of the same class may affect each other during re-transforming. 
Leave this aside for now and add test cases later if incompatibilities arise.

Reply:
> > I think we don't cover multiple plugins for one method case, how could that happens?
> 
> We now support multiple plugins enhancing the same class, although enhancing different methods, but if the same method/field name is generated, there will be an error.
> 
> > And even yoi copied all, it doesn't verify this, because many plugins are not activated(not used) in your scenarios.
> 
> Yes, and we may need more testcases.

@wu-sheng @kylixs  Hello，I wonder if multiple plugins enhancing the same class has been supported?

Reply:
No, it isn't. Also, because we added a new field to the target class, I don't recommend to supporting that.